File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    void generate(List<String> ans,StringBuilder str,int open ,int close,int n){
        if(str.length()==2*n){
            ans.add(str.toString() );
            return;
        }
        if(open<n){
            generate(ans,str.append('('),open+1,close,n);
            str.deleteCharAt(str.length() - 1);

        }
        if(close<open){
            generate(ans,str.append(')'),open ,close+1,n);
            str.deleteCharAt(str.length() - 1);

        }
       
        

    }
    public List<String> generateParenthesis(int n) {
        List<String> ans= new ArrayList<>();
        StringBuilder str=new StringBuilder();
        int open=0;
        int close=0;
        generate(ans,str,open,close,n);
        return ans;

    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void paranthesis(int n,int o,int c,string str,vector<string> & ans){
        if(str.size()==2*n){
            ans.push_back(str);
            return;
        }
        if(o<n){
            paranthesis(n,o+1,c,str+'(',ans);
        }
        if(c<o){
            paranthesis(n,o,c+1,str+')',ans);
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        paranthesis(n,0,0,"",ans);
        return ans;
    }
};